From 7369c8d251793362c64dcbcf37f32fc48c3218b8 Mon Sep 17 00:00:00 2001
From: Herbert Xu <herbert@gondor.apana.org.au>
Date: Mon, 1 Sep 2014 13:10:31 -0400
Subject: [crypto] drbg: DRBG kernel configuration options

Message-id: <E1XORNP-00085N-Gn@gondolin.me.apana.org.au>
Patchwork-id: 90058
O-Subject: [PATCH 3/19] crypto: drbg - DRBG kernel configuration options
Bugzilla: 1011562
RH-Acked-by: Jarod Wilson <jarod@redhat.com>
RH-Acked-by: Rafael Aquini <aquini@redhat.com>

RHEL7 Bugzilla #1011562

Upstream commit: 419090c6c6862abb54b7c77efc3c00ed35909d73

The different DRBG types of CTR, Hash, HMAC can be enabled or disabled
at compile time. At least one DRBG type shall be selected.

The default is the HMAC DRBG as its code base is smallest.

Signed-off-by: Stephan Mueller <smueller@chronox.de>
Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/crypto/Kconfig b/crypto/Kconfig
index 6f8d504..ef552cb 100644
--- a/crypto/Kconfig
+++ b/crypto/Kconfig
@@ -23,7 +23,7 @@ comment "Crypto core or helper"
 
 config CRYPTO_FIPS
  bool "FIPS 200 compliance"
- depends on CRYPTO_ANSI_CPRNG && !CRYPTO_MANAGER_DISABLE_TESTS
+ depends on (CRYPTO_ANSI_CPRNG || CRYTPO_DRBG) && !CRYPTO_MANAGER_DISABLE_TESTS
  depends on MODULE_SIG
  help
    This options enables the fips boot option which is
@@ -1394,6 +1394,40 @@ config CRYPTO_ANSI_CPRNG
    ANSI X9.31 A.2.4. Note that this option must be enabled if
    CRYPTO_FIPS is selected
 
+menuconfig CRYTPO_DRBG
+ tristate "NIST SP800-90A DRBG"
+ depends on CRYPTO
+ select CRYPTO_RNG
+ help
+   NIST SP800-90A compliant DRBG. In the following submenu, one or
+   more of the DRBG types must be selected.
+
+if CRYTPO_DRBG
+
+config CRYPTO_DRBG_HMAC
+ bool "Enable HMAC DRBG"
+ default y
+ depends on CRYTPO_DRBG
+ select CRYPTO_HMAC
+ help
+   Enable the HMAC DRBG variant as defined in NIST SP800-90A.
+
+config CRYPTO_DRBG_HASH
+ bool "Enable Hash DRBG"
+ depends on CRYTPO_DRBG
+ select CRYPTO_HASH
+ help
+   Enable the Hash DRBG variant as defined in NIST SP800-90A.
+
+config CRYPTO_DRBG_CTR
+ bool "Enable CTR DRBG"
+ depends on CRYTPO_DRBG
+ select CRYPTO_AES
+ help
+   Enable the CTR DRBG variant as defined in NIST SP800-90A.
+
+endif #CRYTPO_DRBG
+
 config CRYPTO_USER_API
  tristate
 
-- 
1.7.1